File: python/network/multi_head_attention.py
import torch
import torch.nn as nn
import torch.nn.functional as F


class MultiHeadAttention(nn.Module):
    """
    Computes multi-head attention. Supports nested or padded tensors.
    see: https://docs.pytorch.org/tutorials/intermediate/transformer_building_blocks.html

    Args:
        d_q (int): Size of embedding dim for query
        d_k (int): Size of embedding dim for key
        d_v (int): Size of embedding dim for value
        d_head (int): Size of embedding dim of each head
        nheads (int): Number of heads
        dropout (float, optional): Dropout probability. Default: 0.0
        bias (bool, optional): Whether to add bias to input projection. Default: True
    """

    def __init__(
        self,
        d_q: int,
        d_k: int,
        d_v: int,
        d_head: int,
        nheads: int,
        dropout: float = 0.0,
        bias=True,
        device=None,
        dtype=None,
    ):
        factory_kwargs = {"device": device, "dtype": dtype}
        super().__init__()
        self.d_head = d_head
        self.nheads = nheads
        self.dropout = dropout
        self._qkv_same_embed_dim = d_q == d_k and d_q == d_v
        d_total = d_head * nheads
        if self._qkv_same_embed_dim:
            self.packed_proj = nn.Linear(d_q, d_total * 3, bias=bias, **factory_kwargs)
        else:
            self.q_proj = nn.Linear(d_q, d_total, bias=bias, **factory_kwargs)
            self.k_proj = nn.Linear(d_k, d_total, bias=bias, **factory_kwargs)
            self.v_proj = nn.Linear(d_v, d_total, bias=bias, **factory_kwargs)
        d_out = d_q
        self.out_proj = nn.Linear(d_total, d_out, bias=bias, **factory_kwargs)
        self.bias = bias

    def forward(
        self,
        query: torch.Tensor,
        key: torch.Tensor,
        value: torch.Tensor,
        attn_mask=None,
        is_causal=False,
    ) -> torch.Tensor:
        """
        Forward pass; runs the following process:
            1. Apply input projection
            2. Split heads and prepare for SDPA
            3. Run SDPA
            4. Apply output projection

        Args:
            query (torch.Tensor): query of shape (``N``, ``L_q``, ``d_qk``)
            key (torch.Tensor): key of shape (``N``, ``L_kv``, ``d_qk``)
            value (torch.Tensor): value of shape (``N``, ``L_kv``, ``d_v``)
            attn_mask (torch.Tensor, optional): attention mask of shape (``N``, ``L_q``, ``L_kv``) to pass to SDPA. Default: None
            is_causal (bool, optional): Whether to apply causal mask. Default: False

        Returns:
            attn_output (torch.Tensor): output of shape (N, L_t, E_q)
        """
        # Step 1. Apply input projection
        if self._qkv_same_embed_dim:
            if query is key and key is value:
                result = self.packed_proj(query)
                query, key, value = torch.chunk(result, 3, dim=-1)
            else:
                q_weight, k_weight, v_weight = torch.chunk(
                    self.packed_proj.weight, 3, dim=0
                )
                if self.bias:
                    q_bias, k_bias, v_bias = torch.chunk(
                        self.packed_proj.bias, 3, dim=0
                    )
                else:
                    q_bias, k_bias, v_bias = None, None, None
                query, key, value = (
                    F.linear(query, q_weight, q_bias),
                    F.linear(key, k_weight, k_bias),
                    F.linear(value, v_weight, v_bias),
                )

        else:
            query = self.q_proj(query)
            key = self.k_proj(key)
            value = self.v_proj(value)

        # Step 2. Split heads and prepare for SDPA
        # reshape query, key, value to separate by head
        # (N, L_t, E_total) -> (N, L_t, nheads, E_head) -> (N, nheads, L_t, E_head)
        query = query.unflatten(-1, [self.nheads, self.d_head]).transpose(1, 2)
        # (N, L_s, E_total) -> (N, L_s, nheads, E_head) -> (N, nheads, L_s, E_head)
        key = key.unflatten(-1, [self.nheads, self.d_head]).transpose(1, 2)
        # (N, L_s, E_total) -> (N, L_s, nheads, E_head) -> (N, nheads, L_s, E_head)
        value = value.unflatten(-1, [self.nheads, self.d_head]).transpose(1, 2)

        # Step 3. Run SDPA
        # (N, nheads, L_t, E_head)
        attn_output = F.scaled_dot_product_attention(
            query, key, value, dropout_p=self.dropout, is_causal=is_causal
        )
        # (N, nheads, L_t, E_head) -> (N, L_t, nheads, E_head) -> (N, L_t, E_total)
        attn_output = attn_output.transpose(1, 2).flatten(-2)

        # Step 4. Apply output projection
        # (N, L_t, E_total) -> (N, L_t, E_out)
        attn_output = self.out_proj(attn_output)

        return attn_output

File: cpp/network/src/MultiHeadAttention.cpp
#include "../include/MultiHeadAttention.h"
#include <torch/nn/functional.h>
#include <torch/serialize.h>
#include <torch/torch.h>

MultiHeadAttentionImpl::MultiHeadAttentionImpl(int64_t d_q, int64_t d_k,
                                               int64_t d_v, int64_t d_head,
                                               int64_t nheads, double dropout,
                                               bool bias, torch::Device device,
                                               torch::Dtype dtype) {
  d_head_ = d_head;
  nheads_ = nheads;
  dropout_p_ = dropout;
  use_bias_ = bias;
  qkv_same_embed_dim_ = (d_q == d_k && d_q == d_v);

  const int64_t d_total = d_head * nheads;
  const int64_t d_out = d_q;

  if (qkv_same_embed_dim_) {
    packed_proj_ = register_module(
        "packed_proj",
        torch::nn::Linear(
            torch::nn::LinearOptions(d_q, d_total * 3).bias(bias)));
  } else {
    q_proj_ = register_module(
        "q_proj",
        torch::nn::Linear(torch::nn::LinearOptions(d_q, d_total).bias(bias)));
    k_proj_ = register_module(
        "k_proj",
        torch::nn::Linear(torch::nn::LinearOptions(d_k, d_total).bias(bias)));
    v_proj_ = register_module(
        "v_proj",
        torch::nn::Linear(torch::nn::LinearOptions(d_v, d_total).bias(bias)));
  }

  out_proj_ = register_module(
      "out_proj",
      torch::nn::Linear(torch::nn::LinearOptions(d_total, d_out).bias(bias)));

  dropout_ = register_module("dropout", torch::nn::Dropout(dropout));
  to(device, dtype);
}

torch::Tensor MultiHeadAttentionImpl::forward(
    const torch::Tensor &query, const torch::Tensor &key,
    const torch::Tensor &value, torch::optional<torch::Tensor> attn_mask,
    bool is_causal) {
  torch::Tensor query_projected, key_projected, value_projected;

  // Step 1. Apply input projection
  if (qkv_same_embed_dim_) {
    if (query.data_ptr() == key.data_ptr() &&
        key.data_ptr() == value.data_ptr()) {
      auto result = packed_proj_(query);
      auto chunks = result.chunk(3, -1);
      query_projected = chunks[0];
      key_projected = chunks[1];
      value_projected = chunks[2];
    } else {
      auto weight = packed_proj_->weight;
      auto bias = packed_proj_->bias;
      auto w_chunks = weight.chunk(3, 0);
      if (use_bias_) {
        auto bias_chunks = bias.chunk(3, 0);
        query_projected =
            torch::nn::functional::linear(query, w_chunks[0], bias_chunks[0]);
        key_projected =
            torch::nn::functional::linear(key, w_chunks[1], bias_chunks[1]);
        value_projected =
            torch::nn::functional::linear(value, w_chunks[2], bias_chunks[2]);
      } else {
        query_projected = torch::nn::functional::linear(query, w_chunks[0]);
        key_projected = torch::nn::functional::linear(key, w_chunks[1]);
        value_projected = torch::nn::functional::linear(value, w_chunks[2]);
      }
    }
  } else {
    query_projected = q_proj_(query);
    key_projected = k_proj_(key);
    value_projected = v_proj_(value);
  }

  // Step 2. Split heads: (N, L, E_total) -> (N, nheads, L, d_head)
  query_projected = query_projected
                        .view({query_projected.size(0), query_projected.size(1),
                               nheads_, d_head_})
                        .transpose(1, 2);
  key_projected = key_projected
                      .view({key_projected.size(0), key_projected.size(1),
                             nheads_, d_head_})
                      .transpose(1, 2);
  value_projected = value_projected
                        .view({value_projected.size(0), value_projected.size(1),
                               nheads_, d_head_})
                        .transpose(1, 2);

  // Step 3. Scaled dot-product attention
  auto attn_output = at::native::scaled_dot_product_attention(
      query_projected, key_projected, value_projected, attn_mask, dropout_p_,
      is_causal);

  // (N, nheads, L, d_head) -> (N, L, nheads, d_head) -> (N, L, E_total)
  attn_output = attn_output.transpose(1, 2).reshape(
      {attn_output.size(0), attn_output.size(2), nheads_ * d_head_});

  // Step 4. Output projection
  return out_proj_(attn_output);
}
